docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,757 @@
+tree.MakeDBool(false),       // rolinherit
+tree.MakeDBool(isRoot),      // rolcreaterole
+tree.MakeDBool(isRoot),      // rolcreatedb
+tree.MakeDBool(false),       // rolcatupdate
+tree.MakeDBool(true),        // rolcanlogin
+negOneVal,                   // rolconnlimit
+tree.NewDString("********"), // rolpassword
+tree.DNull,                  // rolvaliduntil
+tree.NewDString("{}"),       // rolconfig
+```
+
+A similar entry must be created for each existing role with the following changes:
+* `isRoot` is replaced with the condition `rolename == admin`
+* `rolcanlogin` is always false
+
+**TODO(mberhault)**: determine if members that inherit superuser roles are listed here with the inherited attributes
+(eg: if `marc ∈ admin`, does `marc` have `rolsuper` set to true?).
+
+#### pg_catalog.pg_auth_members
+
+See [PostgresSQL doc](https://www.postgresql.org/docs/current/static/catalog-pg-auth-members.html)
+
+The new virtual table `pg_auth_members` must be added. It describes role memberships:
+```
+CREATE TABLE pg_catalog.pg_auth_members (
+  roleid           OID,   ; ID of the role that has a member
+  member           OID,   ; ID of the member
+  grantor          OID,   ; ID of the role that granted this membership
+  admin_options    BOOL   ; true if member can grant membership in roleid to others
+)
+```
+
+The `grantor` may or may not be a bogus value.
+
+**TODO(mberhault)**: determine PostgreSQL logic for grantor (especially if original OID is dropped).
+
+#### pg_catalog.pg_authid
+
+See [PostgreSQL doc](https://www.postgresql.org/docs/current/static/catalog-pg-authid.html).
+
+The `pg_authid` virtual table is not currently implemented in Cockroach.
+It contains sensitive information (passwords) and PostgreSQL clearly states:
+```
+Since this catalog contains passwords, it must not be publicly readable. pg_roles is a publicly readable view on pg_authid that blanks out the password field.
+```
+
+We propose to keep in unimplemented, users should query `pg_roles` and `pg_auth_members` instead.
+
+#### information_schema
+
+There are multiple relevant tables in `information_schema` that must be changed or added.
+* `administrable_role_authorizations`: roles the current user has the `ADMIN OPTION` on.
+* `applicable_roles`: roles whose privileges the current user can use.
+* `enabled_roles`: roles the current use is a member of (direct or indirect).
+* `role_table_grants`: similar to the existing `table_privileges`.
+
+**TODO(mberhault)**: determine exactly which are needed and flesh out details.
+
+## Drawbacks
+
+### Cost of lookup
+
+The cost of lookup is currently the biggest issue.
+
+The naive solution must expand the full (or nearly) set of memberships for the incoming user,
+and must do so for every SQL operation.
+
+The [Caching memberships](#caching-memberships) options would make this much more efficient but would
+sacrifice correctness when roles/privileges are changed within a transaction.
```

Comment:
I think we ought to do this. By caching table schema on nodes today we run into the same correctness issues. In particular we support ATOMIC transactions on Schema but not SERIALIZABLE transactions.

Reply:
From a correctness perspective, we absolutely shouldn't, but then again, we shouldn't have privileges in descriptors either.
The caching memberships section is where some of those improvements are proposed, but all of them suck. There may be something we could do w.r.t descriptor leases, but this I where I need some help understanding the options.

Reply:
Schema changes return to a user when they are fully applied. This is done by incrementing the version  on the schema and waiting until there are no outstanding leases on the previous version. I think you can list this as a limitation of the current implementation and do nothing about it for 2.0 . We do not want to support schema changes (version upgrades) on system tables. 

Another thought is to create another table (non system one called role_versions) that can be used by nodes needing a cache of the table data. Caching the system table data requires acquiring a lease on the role_schema table. modifying the system table will also modify the schema of the new table by publishing a new version and waiting for it to be published throughout the cluster. But I feel it's best not done in this RFC.